### PR TITLE
remove external executable dependency 'goimports'

### PR DESCRIPTION
**Pull request by katsusan**
Currently in-use `goimports` command shares the same internal implementation with [golang.org/x/tools/imports](https://golang.org/x/tools/imports). 
But calling an external executable may be unstable and a bit inefficient, so I propose using corresponding library to replace it,
which also solves the problem of building failure in docker as `go get -u golang.org/x/tools/cmd/goimports` needs GOPROXY
to be set if in China mainland.

// Chinese version
英文不太好见谅，大意就是goimports和golang.org/x/tools/imports这个库底层都是调用golang.org/x/tools/internal/imports的
Process函数，因此不会有表现上的不同。此外依赖goimports的话需要在Readme里申明一下prerequisite，并且在docker构建时
如果在大陆网络条件下会go get失败，加入GOPROXY的话也要对国外用户区分，综合以上说法我觉得替换成库实现更简便高效。

### CODE REVIEW

File: src/route/api.go [hunk 1]
```diff
@@ -200,7 +211,7 @@ func autoimports(outf string) error {
 	}
 	return nil
 }
-
+*/
```
Comment (changkun) [inline src/route/api.go]: Cloud you also just delete the above old implementation?

**Review by changkun — COMMENTED** [review]
Very interesting change. Thanks a lot for the information. I didn't know that `goimports` is a direct wrapper of the `golang.org/x/tools/imports`.

- Will the imports package also fix the code formatting?

Could you also clean up a recent change to the command check of `goimports`, see https://github.com/golang-design/ssaplayground/pull/7/files for more details.

This change arises me question regarding getting SSA information more directly, such as:

- Is it possible to directly use https://golang.org/x/tools/go/ssa for generating ssa.html?

**Comment by katsusan** [comment]
-  Will the imports package also fix the code formatting?

Sure. `Process(filename string, src []byte, opt *Options)` provides options 
`opt.FormatOnly` which give you the choice to do format only or not. 
In brief, `opt.FormatOnly == true` → do format only;
`opt.FormatOnly == false` → do autoimport then format.

Looking at following example:

```Golang
package main

import (
	"fmt"

	"golang.org/x/tools/imports"
)

func main() {
	doimports()
}

// source code we want to do autoimport
const src string = `
package main

func main() {
	ctx := context.Background()
	println(ctx)

		r := gin.New()
	println(r)
}
`

func doimports() {
	out, err := imports.Process("", []byte(src), &imports.Options{})
	if err != nil {
		fmt.Println("imports failed,", err)
		return
	}
	fmt.Printf("%s\n", out)
}

```
it will give the output:
```
package main

import (
	"context"

	"github.com/gin-gonic/gin"
)

func main() {
	ctx := context.Background()
	println(ctx)

	r := gin.New()
	println(r)
}
```

- Is it possible to directly use https://golang.org/x/tools/go/ssa for generating ssa.html?

Currently infeasible in my opinion. the generated ssa.html is the human-friendly visualization
of all SSA passes which go team provide, while https://golang.org/x/tools/go/ssa can show 
the presentation of ssa form in a detailed but obscure way (try golang.org/x/tools/cmd/ssadump for example).

In addition, there are other reasons for not doing it :

1. "golang.org/x/tools/go/ssa" is not stable yet.
    `THIS INTERFACE IS EXPERIMENTAL AND IS LIKELY TO CHANGE.` by its document.
2. Go compiler provides options with which you can customize you SSA passes in ssa.html.
    For example, `go build -gcflags=-d=ssa/short_circuit/off` will skip the short_circuit pass.

**Review by changkun — COMMENTED** [review]
It looks like there is a behavior change:

If I use the following snippet:

```
func main() {

	fmt.Println(time.Second)
}
```

Your change will result in:

```
cannot run autoimports for your code, err:
1:1: expected 'package', found 'func'
```

whereas the live server returns ssa.html properly.

**Review by changkun — APPROVED** [review]
Thanks a lot for another awesome fix!

**Comment by katsusan** [comment]
Sorry for not testing fully. New commit has fixed this, which keeps sync with [goimports options](https://github.com/golang/tools/blob/v0.1.0/cmd/goimports/goimports.go#L41-L49)

1. no package declaration with `func main`
2. no package declaration with `func dummy`
3. `package main` + `func main`
4. `package main` + `func dummy`
5.  `package dummy` with `func main`
6. `package dummy` with `func dummy`

Here are test results of  the 6 cases above.

- case1: ok

- case2:   
// same as goimports, the behavior can't be changed,
// see https://github.com/golang/tools/blob/v0.1.0/internal/imports/imports.go#L194-L205
// maybe we can add our package prefix to convert it to case6?
```
cannot build ssa for your code, err:
main$GOSSAPATHgo:2:1: expected 'package', found 'import'
```

- case3: ok

- case4:     
 // same as goimports
```
cannot build ssa for your code, err:
# command-line-arguments
dumped SSA to $GOSSAPATH\ssa$GOSSAPATHhtml
# command-line-arguments
runtime$GOSSAPATHmain_main·f: function main is undeclared in the main package
```

 - case5: ok

- case6: ok

**Pull request merged**